### PR TITLE
Add Action plugin example

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
docs/docsite/rst/dev_guide/developing_plugins.rst

### CODE REVIEW

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 1]
```diff
@@ -74,6 +74,62 @@ Plugins that support embedded documentation (see :ref:`ansible-doc` for the list
 Developing particular plugin types
 ==================================
 
+.. _developing_actions:
+
+Action plugins
+--------------
+
+Action plugins are useful when you want to modify data which is provided to any module or filter the data which is returned by the module.
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
Action plugins are useful when you want to modify data that is provided to any module, or filter the data that is returned by the module.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 2]
```diff
@@ -74,6 +74,62 @@ Plugins that support embedded documentation (see :ref:`ansible-doc` for the list
...
+
+.. code-block:: python
+
+    from ansible.plugins.action import ActionBase
+
+    class ActionModule(ActionBase):
+        pass
+
+
+From there, execute module using ``_execute_module`` method to call original module. After successful execution of module, you can modify module return data.
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
From there, execute the module using the ``_execute_module`` method to call the original module. After successful execution of the module, you can modify the module return data.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 3]
```diff
@@ -74,6 +74,62 @@ Plugins that support embedded documentation (see :ref:`ansible-doc` for the list
...
+
+.. code-block:: python
+
+    module_return = self._execute_module(module_name='<NAME_OF_MODULE>',
+                                         module_args=module_args,
+                                         task_vars=task_vars, tmp=tmp)
+
+The following is an example showing how one can modified return data of Ansible's setup module
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
The following is an example showing how you can modify return data from Ansible's ``setup`` module:
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 4]
```diff
@@ -74,6 +74,62 @@ Plugins that support embedded documentation (see :ref:`ansible-doc` for the list
...
+            ret = dict()
+            if not module_return.get('failed'):
+                for key, value in module_return['ansible_facts'].items():
+                    if key == 'ansible_system':
+                        ret[key] = value
+
+            return dict(ansible_facts=dict(ret))
+
+
+Here, we are capturing ``setup`` module return data and filtering out ``ansible_system`` variable and it's value from the returned data.
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
Here, we are capturing the ``setup`` module return data and filtering out the ``ansible_system`` variable and its value from the returned data.
```

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 5]
```diff
@@ -74,6 +74,62 @@ Plugins that support embedded documentation (see :ref:`ansible-doc` for the list
...
+
+For more example of action plugins, see the source code for the `action plugins included with Ansible Core <https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/action>`_
```
Comment (samccann) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]:
```suggestion
For more examples of action plugins, see the source code for the `action plugins included with Ansible Core <https://github.com/ansible/ansible/tree/devel/lib/ansible/plugins/action>`_
```

**Review by samccann — CHANGES_REQUESTED** [review]

**Comment by acozine** [comment]
With @samccann's suggested edits, this looks like a good addition, thanks @Akasurde! 

Related to, and probably fixes #20557.

File: docs/docsite/rst/dev_guide/developing_plugins.rst [hunk 6]
```diff
@@ -74,6 +74,62 @@ Plugins that support embedded documentation (see :ref:`ansible-doc` for the list
...
+            super(ActionModule, self).run(tmp, task_vars)
+            module_args = self._task.args.copy()
+            module_return = self._execute_module(module_name='setup',
+                                                 module_args=module_args,
+                                                 task_vars=task_vars, tmp=tmp)
+            ret = dict()
+            if not module_return.get('failed'):
+                for key, value in module_return['ansible_facts'].items():
+                    if key == 'ansible_system':
+                        ret[key] = value
```
Comment (bcoca) [inline docs/docsite/rst/dev_guide/developing_plugins.rst]: can i suggest a different example? one that actually shows use of local resources:

timediff action, checks remote clock vs local clock (run date on both machines and compare)

This is not really useful due to network drift but it does show why you would want to use an action plugin.

**Pull request merged**